File: english_learning/tests/territory_catalog.test.js
// Phase 1A frontend tests — canonical territory identity + catalog population.
// Extracts the REAL functions from index.html and the REAL catalog from world-data/.
//   node tests/territory_catalog.test.js
"use strict";
const fs = require("fs");
const path = require("path");
const vm = require("vm");
const assert = require("assert");

const ROOT = path.join(__dirname, "..");
const html = fs.readFileSync(path.join(ROOT, "index.html"), "utf8");

function extractLiteral(src, marker, open) {
  const i = src.indexOf(marker), start = src.indexOf(open, i), close = open === "{" ? "}" : "]";
  let d = 0, k = start;
  for (; k < src.length; k++) { if (src[k] === open) d++; else if (src[k] === close) { d--; if (!d) { k++; break; } } }
  return src.slice(start, k);
}
function extractFn(src, name) {
  const i = src.indexOf("function " + name), start = src.indexOf("{", i);
  let d = 0, k = start;
  for (; k < src.length; k++) { if (src[k] === "{") d++; else if (src[k] === "}") { d--; if (!d) { k++; break; } } }
  return src.slice(i, k);
}
// contiguous catalog block: from `const TERR_CATALOG` up to (not incl) loadTerritoryCatalog
function extractBlock(src, from, to) {
  const a = src.indexOf(from), b = src.indexOf(to, a);
  return src.slice(a, b);
}

const ctx = { console: { warn: function () {} } };
vm.runInNewContext(
  "POP_TABLE = " + extractLiteral(html, "const POP_TABLE", "{") + ";\n" +
  extractFn(html, "popForName") + ";\n" +
  extractBlock(html, "const TERR_CATALOG =", "function loadTerritoryCatalog") + "\n" +
  "this.T = { canonicalTerritoryId, territoryForPath, catalogPopulation, mapIdForSvg, popForName, TERR_CATALOG };",
  ctx
);
const T = ctx.T;

// load the REAL catalog into TERR_CATALOG (mirrors loadTerritoryCatalog without fetch)
const maps = JSON.parse(fs.readFileSync(path.join(ROOT, "world-data", "maps.json"), "utf8"));
maps.forEach(m => { T.TERR_CATALOG.maps[m.id] = m; T.TERR_CATALOG.mapIdBySvg[m.svgFile] = m.id; });
maps.forEach(m => {
  JSON.parse(fs.readFileSync(path.join(ROOT, "world-data", "territories", m.id + ".json"), "utf8"))
    .forEach(t => { T.TERR_CATALOG.terrById[t.id] = t; (t.svgPathKeys || []).forEach(pk => { T.TERR_CATALOG.terrByMapPath[t.mapId + "|" + pk] = t; }); });
});
T.TERR_CATALOG.loaded = true;

let n = 0; function test(name, fn) { fn(); n++; console.log("  ok -", name); }

test("canonical territory ID generation", () => {
  assert.strictEqual(T.canonicalTerritoryId("world", "us"), "world:us");
  assert.strictEqual(T.canonicalTerritoryId("china", "pBJ"), "china:pBJ");
});

test("path resolves to canonical territory (real catalog)", () => {
  assert.strictEqual(T.territoryForPath("china", "pBJ").id, "china:pBJ");
  assert.strictEqual(T.territoryForPath("world", "us").id, "world:us");
});

test("population comes from catalog gamePopulation, not fallback", () => {
  const t = T.territoryForPath("world", "id");                 // Indonesia
  assert.strictEqual(T.catalogPopulation("world", "id", "Indonesia"), t.gamePopulation);
  assert.ok(Number.isInteger(t.gamePopulation) && t.gamePopulation >= 0);
});

test("one logical territory may map to multiple SVG paths", () => {
  // model supports N paths per territory: inject a synthetic multi-path territory
  const terr = { id: "syn:multi", mapId: "syn", regionCode: "multi", gamePopulation: 42, svgPathKeys: ["a", "b", "c"] };
  T.TERR_CATALOG.terrById["syn:multi"] = terr;
  ["a", "b", "c"].forEach(pk => { T.TERR_CATALOG.terrByMapPath["syn|" + pk] = terr; });
  const ra = T.territoryForPath("syn", "a"), rb = T.territoryForPath("syn", "b"), rc = T.territoryForPath("syn", "c");
  assert.ok(ra === rb && rb === rc, "all components resolve to the same territory object");
  assert.strictEqual(ra.id, "syn:multi");
});

test("clicking different components -> same territory & counted once", () => {
  const clickedPaths = ["a", "b", "a", "c"];                    // 4 clicks across 3 components of one territory
  const owned = {};
  clickedPaths.forEach(pk => { const t = T.territoryForPath("syn", pk); owned[t.id] = t.gamePopulation; });
  assert.deepStrictEqual(Object.keys(owned), ["syn:multi"], "one logical territory");
  const totalPop = Object.values(owned).reduce((a, b) => a + b, 0);
  assert.strictEqual(totalPop, 42, "population counted once, not per component");
});

test("missing catalog entry uses controlled fallback (popForName)", () => {
  const got = T.catalogPopulation("world", "zzz", "Neverland");
  assert.strictEqual(got, T.popForName("Neverland"), "falls back deterministically");
});

console.log("\nAll " + n + " frontend catalog tests passed.");

File: english_learning/tests/territory_catalog_test.py
#!/usr/bin/env python3
"""Phase 1A backend tests — catalog load, validation, legacy resolution, freeze.

    python3 tests/territory_catalog_test.py
"""
import json
import os
import re
import sys
import tempfile

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.insert(0, ROOT)
from territory_catalog import TerritoryCatalog  # noqa: E402

passed = 0


def ok(name):
    global passed
    passed += 1
    print("  ok -", name)


def make_catalog(maps, terrs):
    """Build a throwaway catalog dir and return a loaded TerritoryCatalog."""
    d = tempfile.mkdtemp()
    os.makedirs(os.path.join(d, "territories"), exist_ok=True)
    json.dump(maps, open(os.path.join(d, "maps.json"), "w"))
    for mid, ts in terrs.items():
        json.dump(ts, open(os.path.join(d, "territories", mid + ".json"), "w"))
    return TerritoryCatalog(base_dir=d).load()


def terr(tid, mid, code, pop=100, keys=None):
    return {"id": tid, "mapId": mid, "regionCode": code, "displayName": code,
            "localizedNames": {"en": code}, "svgPathKeys": keys or [code],
            "administrativeCode": None, "gamePopulation": pop,
            "populationSource": {"value": None, "year": None, "sourceName": None, "sourceUrl": None},
            "metadata": {}, "terrainType": None, "settlementType": None,
            "features": [], "adjacentTerritoryIds": []}


# --- 1. real catalog loads and validates ---
real = TerritoryCatalog().load()
assert real.validate() == [], real.validate()
counts = real.count_per_map()
assert counts.get("world", 0) > 0 and counts.get("china", 0) > 0, counts
ok("real catalog loads + validates (%s)" % counts)

# --- 2. duplicate territory id fails ---
c = make_catalog([{"id": "m", "svgFile": "maps/x.svg", "name": "M"}],
                 {"m": [terr("m:a", "m", "a"), terr("m:a", "m", "a")]})
assert any("duplicate territory id" in e for e in c.validate())
ok("duplicate territory id fails")

# --- 3. duplicate map id fails ---
c = make_catalog([{"id": "m", "svgFile": "a", "name": "M"}, {"id": "m", "svgFile": "b", "name": "M2"}],
                 {"m": [terr("m:a", "m", "a")]})
assert any("duplicate map id" in e for e in c.validate())
ok("duplicate map id fails")

# --- 4. invalid map reference fails ---
c = make_catalog([{"id": "m", "svgFile": "a", "name": "M"}],
                 {"m": [terr("x:a", "x", "a")]})
assert any("unknown map" in e for e in c.validate())
ok("invalid map reference fails")

# --- 5. negative population fails ---
c = make_catalog([{"id": "m", "svgFile": "a", "name": "M"}],
                 {"m": [terr("m:a", "m", "a", pop=-5)]})
assert any("bad gamePopulation" in e for e in c.validate())
ok("negative population fails")

# --- 6/7. legacy world + china keys resolve ---
assert real.canonical_from_legacy("maps/world.svg#us") == "world:us"
assert real.canonical_from_legacy("maps/china.svg#pBJ") == "china:pBJ"
ok("legacy world + china keys resolve to canonical ids")

# --- 8. unknown legacy key handled safely ---
assert real.canonical_from_legacy("maps/world.svg#zzz") is None
assert real.canonical_from_legacy("garbage") is None
ok("unknown legacy key returns None (no crash, not discarded silently)")

# --- 9. multi-path region resolves to one logical territory ---
c = make_catalog([{"id": "m", "svgFile": "maps/m.svg", "name": "M"}],
                 {"m": [terr("m:big", "m", "big", keys=["p1", "p2", "p3"])]})
ids = {c.canonical_from_legacy("maps/m.svg#" + p) for p in ("p1", "p2", "p3")}
assert ids == {"m:big"}, ids
ok("multi-path region resolves to ONE territory (p1/p2/p3 -> m:big)")

# --- 10. population lookup uses gamePopulation ---
assert real.game_population("china:pBJ") == real.territories["china:pBJ"]["gamePopulation"]
ok("population lookup returns gamePopulation")

# --- 11. deterministic legacy fallback value is frozen correctly ---
# Reproduce index.html popForName's hash (no POP_TABLE hit for China names) in Python
# and confirm the catalog froze that exact value.
def js_pop_hash(name):
    h = 0
    for ch in name:
        h = (h * 31 + ord(ch)) & 0xFFFFFFFF
    return 60 + (h % 241)

bj = real.territories["china:pBJ"]
assert bj["displayName"] == "Beijing 北京", bj["displayName"]
assert bj["gamePopulation"] == js_pop_hash("Beijing 北京"), (bj["gamePopulation"], js_pop_hash("Beijing 北京"))
ok("deterministic hash population frozen correctly (china:pBJ = %d)" % bj["gamePopulation"])

print("\nAll %d backend catalog tests passed." % passed)
